### DISCUSSION TITLE
How to find the "bad" bookmark that leads to a sync failure in Firefox with Nextcloud Bookmarks?

### DISCUSSION THREAD

**Discussion by johanneskastl** [floccusaddon/floccus#1760] [Q&A] [answered]
Hi all,

is there a way to find out, which of my bookmarks is causing the sync to fail?

I am syncing from Firefox to my Nextcloud Bookmarks. The sync starts but fails with this error:

```
["Could not add bookmark"]} Nextcloud API error: {"status":"error","data":["Could not add bookmark"]} Nextcloud API error: {"status":"error","data":["Could not add bookmark"]} | Last synchronized: 0 seconds ago
```

I create a backup of my bookmarks, and delete half of them. Now the sync is successful.

So, other than importing the backup, only deleting half of the half that I deleted previously, and repeating this for ever, how can I find out which of the bookmarks is causing the sync to fail?

The debug log looks like this, it does not indicate which bookmark failed:

```
2024-11-08T07:31:02.387Z Executed 8 actions from 426 actions
2024-11-08T07:31:02.388Z Executed 9 actions from 426 actions
2024-11-08T07:31:02.463Z Receiving response for PUT https://<redacted>/index.php/apps/bookmarks/public/rest/v2/bookmark/71
2024-11-08T07:31:02.465Z Executed 10 actions from 426 actions
2024-11-08T07:31:02.465Z Syncing failed with Nextcloud API error:
{"status":"error","data":["Could not add bookmark"]}
Nextcloud API error:
{"status":"error","data":["Could not add bookmark"]}

```

I already skimmed for "funny" characters in the names and renamed the obvious ones. But to no avail (and I have a lot of bookmarks...).

Thanks in advance!

Kind Regards,
Johannes

**Comment by marcelklehr** [comment]
Hey Johannes,

the PUT should be associated with an UPDATE action, which should contain the Bookmark that is being updated in the log entry. Since this seems to happen in the first 10 actions being executed, you should be able to find the offending Bookmark in the logs leading up to this point. If you like you can send me the log privately via email.

**Reply by johanneskastl** [reply]
Thanks for the reply, Marcel. My problem with the logs is that there seem to be multiple lines for each bookmark, but not "grouped" by each transaction. Instead they appear interweaved.

But I'll have a closer look and see if I can find out which one fails.

Is there a way to find out why it could not be added? Funny characters, it already exists, the folder does not exist, I do not know what else could go wrong...

**Reply by johanneskastl** [reply]
OK, I think I got the relations right. But I find no reason, why such a Bookmark should cause issues?

This is the line for the offending bookmark:

```
           - #1340;31[Kitsch trifft Armut](http://www.spiegel.de/kultur/gesellschaft/0,1518,714571,00.html) parentId: 31
```

**Reply by marcelklehr** [reply]
Mmmh, the IDs are not the same, though. 1340 is the id of the last bookmark, 71 the one of the offending bookmark

**Reply by marcelklehr** [reply]
Looking deeper into this, though, the API error stems from this method: https://github.com/nextcloud/bookmarks/blob/484d6c26ea8edcb590572d88dccbdf8c826c32ba/lib/Controller/BookmarkController.php#L427-L457 which means it can't have stemmed from the PUT 71, it must have been from a CREATE action, which would be a POST request

**Reply by johanneskastl** [reply]
> Mmmh, the IDs are not the same, though. 1340 is the id of the last bookmark, 71 the one of the offending bookmark

Sorry, I took a newer debug log and use that as an example.

There  are at least 100 lines between the last CREATE.

The last PUT is a dozen lines before:

```
2024-11-09T18:01:04.203Z Receiving response for POST https://<redacted>/index.php/apps/bookmarks/public/rest/v2/bookmark
```

This does not shed much light, unfortunately...

**Reply by marcelklehr** [reply]
Can you send me a redacted log perhaps? That would make things easier

**Comment by marcelklehr** [comment]
I was able to replicate this issue, it doesn't seem to be due to the URL. Can you try emptying the trash in Nextcloud Bookmarks?

**Reply by johanneskastl** [reply]
Thanks for digging into this. I'll empty the trash, sure.

**Reply by johanneskastl** [reply]
I move some bookmarks into one folder that is being synced, but after some folders I ran into issues again. I'll have to cleanup the lots of duplicates before I can proceed, but this way I noticed once there are errors.

I'll clean up the mess and get a new debug log, once I can nail it down.

**Reply by marcelklehr** [reply]
From what I found, this may indeed be due to duplicates. The bookmarks app tries to manage duplicates in a smart way, by having multiple references to the same bookmark when the same URL is used in multiple places, but when a bookmark is edited to have the URL of a bookmark that already exists, this system breaks down currently and there are two actual duplicate bookmarks, which can cause issues when handling these.

**Reply by johanneskastl** [reply]
Hmmm, in my case I got lots of folders being duplicated. But not in the same way each time.

One folder I got five times, four of them being emtpy, only one having bookmarks in it.
Another folder I also got 5 times, but all of them were empty.

I am currently cleaning up and will try to get the sync working again...

**Reply by marcelklehr** [reply]
Mh, do you know how you got these duplicates. Duplicates can result from using the browser built-in sync in parallel to floccus, so best check that you have e.g. Firefox Sync disabled.

**Reply by johanneskastl** [reply]
I do not have Firefox Sync enabled. Those seem to come from my experiments with gradually moving more of my bookmarks into the `Other Bookmarks` folder that is being synced to Nextcloud using the Floccus addon.

**Reply by johanneskastl** [reply]
See https://github.com/floccusaddon/floccus/discussions/1764

**Reply by marcelklehr** [reply]
Ah, that one slipped through while I was moving house. Sorry.

**Reply by johanneskastl** [reply]
No problem, I am not sure if it is related to the "bad bookmark" issue or not, so I pointed it out.

Basically I would have assumed that there can only be one folder with a certain name, but apparently there can be multiple. Which confuses me and might confuse Floccus and/or Nextcloud Bookmarks?

**Reply by marcelklehr** [reply]
It can confuse floccus in certain situations, yes. The outcome would likely be a floccus error, though, not a Nextcloud Bookmarks error as in your case. :thinking:

**Reply by johanneskastl** [reply]
OK, I noticed something.

I am gradually moving folders from the Bookmark menu (which is not synced) to `Other Bookmarks`, which is being synced.

Some of the folders cause an error. Upon retry I find duplicate folders in Nextcloud Bookmarks, apparently one for each failed sync. Moving the folder back (in Firefox), removing all of the "duplicate folders" and then re-trying the sync normally succeeds. Of course that is tedious.

I also noticed that sometimes a folder is being moved, the sync is successful, but the next sync (or one of the next syncs)  errors out. Not sure why that is.

**Reply by marcelklehr** [reply]
Do you have an error message for me?

**Reply by johanneskastl** [reply]
I am trying to nail it down to a "bad" bookmark and will send a log to you.

**Reply by johanneskastl** [reply]
OK, one thing I found:

Bookmarks with names that exceed a certain number of characters cause errors.

Failing:
```
Enigmail / Bugs / #436 TB error "Unable to save your message as draft." when drafts are encrypted; Enigmail error: "mimeEncrypt.js: caught exception: undefined"
```

Succeeding:
```
Enigmail / Bugs / #436 TB error "Unable to save your message as draft." when drafts are encrypted; Enigmail error: 
```

**Reply by marcelklehr** [reply]
I believe this was fixed in Nextcloud Bookmarks a while ago to automatically truncate the title. Which version do you have?

**Reply by johanneskastl** [reply]
Nextcloud says the Bookmarks app is at 14.2.6

**Reply by johanneskastl** [reply]
OK, shortening the title to something roughly around 117 characters seems to avoid the errors.

**Reply by marcelklehr** [reply]
117? Then your DB schema isn't up to date, I think.

**Reply by marcelklehr** [reply]
Is it possible that you had installed the bookmarks app once years ago, uninstalled it and now installed it again? You may be running an up-to-date code base on an outdated DB schema. The up-to-date schema should allow 1000 bytes in the title.

**Reply by johanneskastl** [reply]
> Is it possible that you had installed the bookmarks app once years ago, uninstalled it and now installed it again? You may be running an up-to-date code base on an outdated DB schema. The up-to-date schema should allow 1000 bytes in the title.

Not willingly, but as sometimes apps get deactivated during updates, I cannot rule that out.

Is there a way to update the schema?

**Reply by marcelklehr** [reply, marked as answer]
First please check the oc_bookmarks table schema and post it here, so we can check if it's actually outdated.

Then the easiest way is to

0. Put your instance in maintenance mode
1. make a db backup
2. log into your dbms console
3. drop all tables that start with oc_bookmarks
4. Delete all entries in oc_migrations for the bookmarks app
5. Run occ upgrade

**This will clear all bookmarks data for all users on your instance, though.** It's essentially a reinstall. If you want to keep your data, you need to manually adjust the db tables. You need to check the Migrations folder in apps/bookmarks/lib/Migrations and figure out from the code which changes to make to the db schema. This is difficult though, because you don't know which schema changes are missing.

**Reply by johanneskastl** [reply]
As I am just starting to use Floccus and Nextcloud (as the first user on this instance), I can start from scratch without issues.

Is there an easy way to uninstall Nextcloud bookmarks and remove the database tables? To make sure I do not have any leftovers in the database?

Or do I need to uninstall and drop the tables manually?

**Reply by marcelklehr** [reply]
> Is there an easy way to uninstall Nextcloud bookmarks and remove the database tables?

The easy way is outlined above, sadly.

**Reply by johanneskastl** [reply]
OK, I'll give it a try and report back.

**Reply by johanneskastl** [reply]
Short question:

`Delete all entries in oc_migrations for the bookmarks app`

Is this a table called `oc_migrations` or is this a folder in config or similar?

**Reply by marcelklehr** [reply]
yes it's a table in the db

**Reply by johanneskastl** [reply]
Forget my last question, I just overlooked the table `oc_migrations`.

I completed the steps, but the `occ upgrade` only said that it is already the latest version.

**Reply by marcelklehr** [reply]
Mh, can you try disabling the bookmarks app and enabling it again?

**Reply by marcelklehr** [reply]
because we removed the migrations entries, nextcloud should now realize that it needs to run migrations for the bookmarks app

**Reply by johanneskastl** [reply]
I had to deactive and remove the Bookmarks app, otherwise the /apps/bookmarks URL spit out a server error.

Now it seems to be working again, I'll start the syncs again and report back, if I get a smoother ride now and more than 117 characters... :-)

**Reply by johanneskastl** [reply]
Thanks a lot for your help, Marcel!

**Reply by johanneskastl** [reply]
OK, seems like the 117 character limit is no longer a problem. Hooray!

**Reply by johanneskastl** [reply]
Hmm, could the outdated db_schema also explain why I seem to have lost keywords on some bookmarks? Some are still working, but others just vanished apparently.

**Reply by marcelklehr** [reply]
Floccus doesn't sync keywords / tags, so you likely lost those when droppign the tables

**Reply by johanneskastl** [reply]
I noticed them during the lots of sync issues I had, even before I dropped the tables and uninstalled/reinstalled the bookmarks app today.

I'll have a closer look on it and see, if it happens again.

**Reply by johanneskastl** [reply]
Oh, and I mean "Schlüsselwörter", not "Schlagwörter". No idea, how exactly those are translated in English...

**Reply by marcelklehr** [reply]
Ah, floccus can't sync those either :(

**Reply by johanneskastl** [reply]
OK, then it might be that some of the bookmarks where re-created on one device from Floccus/NC Bookmarks, while on another one it was the "original" one with the keyword.
